test/pthread/test_pthread_proxy_deadlock.c

```diff
@@ -0,0 +1,47 @@
+// Copyright 2025 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+#include <pthread.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <emscripten/console.h>
+#include <emscripten/heap.h>
+#include <emscripten/proxying.h>
+#include <emscripten/threading.h>
+
+bool should_quit = false;
+pthread_t looper;
+
+// In the actual implementation of malloc the system queue may be executed
+// non-deterministically if malloc is waiting on a mutex. This wraps malloc and
+// executes the system queue during every allocation to make the behavior
+// deterministic.
+void *malloc(size_t size) {
+  if (emscripten_proxy_get_system_queue() && emscripten_is_main_runtime_thread()) {
+    emscripten_proxy_execute_queue(emscripten_proxy_get_system_queue());
+  }
+  void *ptr = emscripten_builtin_malloc(size);
+  return ptr;
+}
+
+void run_on_looper(void* arg) {
+  emscripten_out("run_on_looper\n");
+  should_quit = true;
+}
+
+void* looper_main(void* arg) {
+  while (!should_quit) {
+    emscripten_proxy_execute_queue(emscripten_proxy_get_system_queue());
+    sched_yield();
+  }
+  return NULL;
+}
+
+int main() {
+  pthread_create(&looper, NULL, looper_main, NULL);
+  emscripten_proxy_async(emscripten_proxy_get_system_queue(), looper, run_on_looper, NULL);
```

Comment:
Could this ever set should_quit before looper_main actually starts running?

Can you add some comments somewhere in this file about exactly that case we are testing for.

Reply:
I kind of assumed the thread start would be guaranteed to run before any proxy messages, but now I'm not sure. Even if it does run before, we'd still hit a deadlock since the main thread will be locked during task creation and trying to run the queue and lock again. It doesn't really matter what the other thread is doing.

Reply:
You can have `looper_main` set a flag and have `main` wait for that flag to be set before doing the proxying to be sure. +1 to adding comments, especially about how the deadlock would occur without the bug fix.

Reply:
I changed the test a bit more and removed the looper thread and just sends the task to the same thread. This makes the test a little less realistic, but highlights it doesn't really matter what the thread is doing.